include/sys/onoff.h

```
@@ -0,0 +1,606 @@
+		} callback;
+	} async;
+
+	/*
+	 * The result of the operation.
+	 *
+	 * This is the value that was (or would be) passed to the
+	 * async infrastructure.  This field is the sole record of
+	 * success or failure for no-wait synchronous operations.
+	 */
+	int volatile result;
+
+	/* Flags recording client state. */
+	unsigned int volatile flags;
+};
+
+/**
+ * @brief Check for and read the result of an asynchronous operation.
+ *
+ * @param op pointer to the object used to specify asynchronous
+ * function behavior and store completion information.
+ *
+ * @param result pointer to storage for the result of the operation.
+ * The result is stored only if the operation has completed.
+ *
+ * @retval 0 if the operation has completed.
+ * @retval -EAGAIN if the operation has not completed.
+ */
+static inline int onoff_client_fetch_result(const struct onoff_client *op,
+					    int *result)
+{
+	__ASSERT_NO_MSG(op != NULL);
+	__ASSERT_NO_MSG(result != NULL);
+
+	int rv = -EAGAIN;
+
+	if (op->flags == 0U) {
+		rv = 0;
+		*result = op->result;
+	}
+	return rv;
+}
+
+/**
+ * @brief Initialize an on-off client to be used for a spin-wait
+ * operation notification.
+ *
+ * Clients that use this initialization receive no asynchronous
+ * notification, and instead must periodically check for completion
+ * using onoff_client_fetch_result().
+ *
+ * @param cli pointer to the client state object.
+ */
+static inline void onoff_client_init_spinwait(struct onoff_client *cli)
+{
+	__ASSERT_NO_MSG(cli != NULL);
+
+	*cli = (struct onoff_client){
+		.flags = ONOFF_CLIENT_NOTIFY_SPINWAIT,
+	};
+}
+
+/**
+ * @brief Initialize an on-off client to be used for a signal
+ * operation notification.
+ *
+ * Clients that use this initialization will be notified of the
+ * completion of operations submitted through onoff_request() and
+ * onoff_release() through the provided signal.
+ *
+ * @note
+ *   @rst
+ *   This capability is available only when :option:`CONFIG_POLL` is
```

@nashif  with the DOXYGEN macro, calling this with `CONFIG_POLL` unset is a nice, fail-fast build error. Without it, it's accepted by the compiler but silently does nothing.

Do you really think that is better?

---

Note that later changes made by @pabigot make this a runtime error rather a build-time error. So still errors out, but still not really as nice IMHO. I won't get in the way further however.

---

IIRC when I originally checked this I got a link-time error because the poll functions that the application used were not built in.  The changes for the runtime error cover a situation where the application doesn't require the poll functions, but still configures for a signal notification.

The goal of the later changes was to eliminate a path where there was a silent failure.